### PR TITLE
materialized: absorb coordd

### PR DESCRIPTION
**Pull request by benesch**
To run the demo:

```
# Turn on a storage process.
cargo run --bin dataflowd -- --runtime=storage --listen-addr=localhost:6876 --storage-addr=localhost:6700
# Turn on a compute process.
cargo run --bin dataflowd -- --runtime=compute --listen-addr=localhost:6877 --storage-addr=localhost:6700 --storage-workers=1
# Turn on materialized.
cargo run -- --dev -w1 --experimental --storage-compute-addr=localhost:6700 --storage-controller-addr=localhost:6876 --storage-workers=1
# Connect via psql.
psql -h localhost -p 6875 -U materialize materialize
# Create and use a remote cluster.
CREATE CLUSTER c REMOTE 'localhost:6877';
SET CLUSTER = c;
CREATE MATERIALIZED VIEW v AS SELECT 1;
SELECT * FROM v;
1
# Kill the compute process and watch everything explode.
```

----

Teach `materialized` to absorb all the responsibilities of `coordd`, and
then delete `coordd`. This is sort of like #11132, except less stressful
from a security point of view, as this commit is careful not to open any
new ports in `materialized`.

`materialized` always hosts a single compute runtime to host virtual
clusters. All vitual clusters are created in this in-process runtime.

By default, `materialized` also hosts a storage runtime in the same
timely cluster as the compute runtime.  Optionally, the new
storage command-line flags

  * --storage-workers
  * --storage-compute-addr
  * --storage-controller-addr

can be passed to specify that `materialized` should not start a storage
timely cluster, but instead connect to the storage process.

`CREATE CLUSTER` also learns to take a new `REMOTE` option, so that it
is possible to test connecting to a cluster that is hosted in a
different process. This option will never be exposed to users, but it is
useful for testing now.

The `LegacyMultiplex` mode is removed from `dataflowd`, as it was no
longer (easily) possible to use it with `materialized`, as
`materialized` would try to connect twice: once via the storage client
and once for the compute client. The philosophy is that either we run in
single-process mode or fully-distributed mode; there's not much need to
support a semi-distributed mode. The cluster tests are updated to only
exercise the fully-distriburted mode.

A large part of this commit is refactoring the `Controller`/`Client`
stack so that the `Controller` manages separate storage clients and
compute clients, with one compute client per compute instance.
Previously, the `SplitClient` type would snoop commands as they flowed
past and manufacture new clients on the fly. The new approach makes the
structure of the `Controller` more logical, at the cost of some
complicated async code that multiplexes commands and demultiplexes
responses. This code will vanish, though, if we eventually get rid of
virtual clusters.

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

Not sure how much of this is mergeable, but it makes for a good demo.

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - N/A

### CODE REVIEW

File: src/dataflowd/src/lib.rs [hunk 1]
```diff
@@ -8,245 +8,3 @@
...
-    pub fn framed_client<C>(conn: C) -> FramedClient<C>
-    where
-        C: AsyncRead + AsyncWrite,
-    {
-        tokio_serde::Framed::new(
-            tokio_util::codec::Framed::new(conn, length_delimited_codec()),
-            Bincode::default(),
-        )
-    }
-}
```
Comment (benesch) [inline src/dataflowd/src/lib.rs:252]: This got moved wholesale into `dataflow_types::client`.

File: src/dataflow-types/src/client.rs [hunk 2]
```diff
@@ -964,3 +1122,248 @@ pub mod process_local {
...
+            self.connection
+                .next()
+                .await
+                .map(|x| x.expect("connection to dataflow server broken"))
+        }
+    }
+
+    /// A framed connection to a dataflowd server.
+    pub type Framed<C, T, U> = tokio_serde::Framed<
+        tokio_util::codec::Framed<C, LengthDelimitedCodec>,
```
Comment (frankmcsherry) [inline src/dataflow-types/src/client.rs:1173]: My eyes glazed over reading this. Is there a good way to document what the type parameters mean? Alternately, longer names to use for them?

File: src/dataflow-types/src/client.rs [hunk 3]
```diff
@@ -964,3 +1122,248 @@ pub mod process_local {
...
+        async fn recv(&mut self) -> Option<Response<T>> {
+            // TODO: something better than panicking.
+            self.connection
+                .next()
+                .await
+                .map(|x| x.expect("connection to dataflow server broken"))
+        }
+    }
+
+    /// A framed connection to a dataflowd server.
```
Comment (frankmcsherry) [inline src/dataflow-types/src/client.rs:1171]: I think we could spruce this up a bit, maybe not immediately, but .. a "`dataflowd` server" seems like a weird concept if it is a compute instance, as it serves nothing in that case and is only a client.

File: src/dataflow-types/src/client.rs [hunk 4]
```diff
@@ -964,3 +1122,248 @@ pub mod process_local {
...
+    use crate::client::GenericClient;
+
+    /// A client composed of channels.
+    #[derive(Derivative)]
+    #[derivative(Debug)]
+    pub struct ChannelClient<C1, C2, R> {
+        tx: UnboundedSender<C2>,
+        rx: UnboundedReceiver<R>,
+        #[derivative(Debug = "ignore")]
+        command_map: Box<dyn Fn(C1) -> C2 + Send>,
```
Comment (frankmcsherry) [inline src/dataflow-types/src/client.rs:1237]: Nit, but could this be `FnMut`? I'm thinking of all the `Client` implementations I've written and they are largely stateful. Maybe that is not what this is for though.

**Review by frankmcsherry — COMMENTED** [review]
I've read everything and I think with an explanation of some parts, I could imagine it making sense and being good to land. I think I need to get a better handle on the control flow with the virtual compute hosts, though. It seems plausibly great, but I can't yet reverse its behavior out of the implementation with confidence.

Also, `test/cluster/mzcompose.py` drops a lot of cluster tests. I couldn't tell if they survive somewhere else now, or if it is just that with `coordd` gone they needed to go too. If they are replace, great! And if not, we should discuss getting replacements in place if possible!

File: src/dataflow/src/server.rs [hunk 5]
```diff
@@ -91,7 +91,9 @@ pub struct Server {
 /// Initiates a timely dataflow computation, processing materialized commands.
 ///
 /// It uses the default [EventLinkBoundary] to host both compute and storage dataflows.
-pub fn serve(config: Config) -> Result<(Server, BoundaryHook<LocalClient>), anyhow::Error> {
+pub fn serve(
+    config: Config,
+) -> Result<(Server, BoundaryHook<LocalClient<mz_repr::Timestamp>>), anyhow::Error> {
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/dataflow/src/server.rs]: Fixed!

File: src/dataflow-types/src/client.rs [hunk 6]
```diff
@@ -410,36 +413,53 @@ pub trait Client<T = mz_repr::Timestamp>: std::fmt::Debug {
...
+    async fn recv(&mut self) -> Option<R> {
         (**self).recv().await
     }
 }
 
 /// A convenience type for compatibility.
 #[derive(Debug)]
-pub struct LocalClient {
-    client: partitioned::Partitioned<process_local::ProcessLocal>,
+pub struct LocalClient<T> {
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/dataflow-types/src/client.rs]: Done as well!

File: src/dataflow-types/src/client.rs [hunk 7]
```diff
@@ -77,9 +80,9 @@ pub enum ComputeCommand<T = mz_repr::Timestamp> {
     /// Indicates the creation of an instance, and is the first command for its compute instance.
     ///
     /// Optionally, request that the logging sources in the contained configuration are installed.
-    CreateInstance(InstanceConfig, Option<LoggingConfig>),
+    InitializeInstance(Option<LoggingConfig>),
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/dataflow-types/src/client.rs]: (For posterity: this was a change from an earlier iteration of a branch where virtual compute clusters became local compute clusters, and so they were more "initialized" than "created". Restored the original name.)

File: src/dataflow-types/src/client.rs [hunk 8]
```diff
@@ -77,9 +80,9 @@ pub enum ComputeCommand<T = mz_repr::Timestamp> {
...
     /// Indicates the termination of an instance, and is the last command for its compute instance.
-    DropInstance,
+    DeinitializeInstance,
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/dataflow-types/src/client.rs]: Ditto

File: src/dataflow-types/src/client.rs [hunk 9]
```diff
@@ -964,3 +1122,248 @@ pub mod process_local {
...
+                },
+                Some(cmd) = command_rx.recv() => client.send(cmd).await.unwrap(),
+                Some(res) = client.recv() => match res {
+                    Response::Storage(res) => storage_response_tx.send(res).unwrap(),
+                    Response::Compute(res, instance) => match instances.get(&instance) {
+                        None => {
+                            error!("received response {res:?} for non-existent compute instance {instance:?}");
+                        }
+                        Some(response_tx) => response_tx.send(res).unwrap(),
+                    },
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/dataflow-types/src/client.rs]: Yeah, good call. Fixed with a `biased` in the macro.

**Review by frankmcsherry — APPROVED** [review]
Ah, this looks good to me now!

**Comment by benesch** [comment]
Into the abyss we go!

**Pull request merged**